compiler/analyzer.ml

```diff
@@ -470,6 +518,41 @@ and check_func_call_args env id f args =
     env', args'
   else env', args'
 
+and check_func_call_ret env id args ret_default =
+  let id' = match id with
+    | Ast.Id(id') -> id'
+    | _ -> raise (Semantic_Error "check_func_call_ret() called with non-ID") in
+  let builtin =
+    if VarMap.mem id' env.scope then
+      (VarMap.find id' env.scope).builtin
+    else false in
+  if not builtin then env, ret_default else
+  
+  match id' with
+    | "head" -> let Sast.Expr(_, typ) = List.hd args in
+        begin match typ with
+          | List(t) -> env, t
+          | _ -> env, ret_default (* NEVER SHOULD BE REACHED TO DO: MAKE ERROR *)
+        end
+    | "tail" -> let Sast.Expr(_, typ) = List.hd args in env, typ
+    | "cons" ->
+        let Sast.Expr(cons, c_typ) = List.hd args and
+          Sast.Expr(l, l_typ) = List.hd (List.tl args) in
+        let l_elem_typ = begin match l_typ with
+          | List(t) -> t
+          | _ -> ret_default (* NEVER SHOULD BE REACHED TO DO: MAKE ERROR *)
```

Comment:
want to do this in this PR? we can have a deadcodepath_error or something, there's probably one or two other places it could be used